### PR TITLE
feat(ts-sinon): implementing ts-sinon package

### LINKED ISSUE #604: Implementing `ts-sinon`, based on `ts-jest` functionality
### Description
This issue is to outline the development plan for the implementation of `ts-sinon`. The primary goal of this package is to aid developers in creating type-safe, comprehensive, and versatile mock objects for their unit tests. `ts-sinon` will be based on the features and ease of use offered by `ts-jest`.

### Features
- A `createMock` function to generate fully mocked objects based on their types

### PR DESCRIPTION
**Pull request by omermorad**
## Description

This pull request introduces the primary implementation of the `ts-sinon` package. Inspired by the features and ease of use offered by `ts-jest`, `ts-sinon` is designed to facilitate a wide array of testing scenarios and requirements.

Through this pull request, our goal is to equip developers with a powerful and efficient mocking tool that seamlessly integrates into their testing workflows, enhancing the robustness and reliability of their tests.

Closes #604

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
   "license": "MIT",
   "workspaces": [
     "packages/*",
+    "packages/*/*",
```
Comment (underfisk) [inline package.json]: Does line `6` still apply if you're?

File: packages/testing/ts-sinon/src/mocks.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import * as sinon from 'sinon';
+import { SinonStub } from 'sinon';
+
+type DeepPartial<T> = {
```
Comment (underfisk) [inline packages/testing/ts-sinon/src/mocks.ts:4]: Were the types/helpers copied from `ts-jest`? If so i would recommend having a common package to re-usable the code/types between code

**Review by underfisk — CHANGES_REQUESTED** [review]
lgtm, I'm just concern on duplicating code/types but the feature looks pretty handy for sinon users

**Comment by underfisk** [comment]
@omermorad Could you merge the issues and address feedback? If so I'll be willing to push this PR

File: packages/testing/ts-sinon/src/mocks.ts (hunk 2 above)
Comment (omermorad) [reply packages/testing/ts-sinon/src/mocks.ts:4]: So you want to remove the types from `ts-jest` and `ts-sinon` and put them in `@golevelup@tests-types`?

**Comment by omermorad** [comment]
> @omermorad Could you merge the issues and address feedback? If so I'll be willing to push this PR

@underfisk Sure, I apologize for the delay. I'll have time this weekend to fix it; in the meantime, I responded to your comment.

File: packages/testing/ts-sinon/src/mocks.ts (hunk 2 above)
Comment (underfisk) [reply packages/testing/ts-sinon/src/mocks.ts:4]: @omermorad If the implementation is the same but what differs is the underlying types/package we could extract it to a common place. I don't think it should be a blocker though but could be a follow up

Comment (omermorad) [reply packages/testing/ts-sinon/src/mocks.ts:4]: So, shall we merge this pull request and open a new issue for further discussion?

Comment (underfisk) [reply packages/testing/ts-sinon/src/mocks.ts:4]: @omermorad Sure, I'm down for it :)

**Review by underfisk — APPROVED** [review]
lgtm, as soon as you merge the conflicts and CI passes I'll try to push this

**Review by underfisk — APPROVED** [review]

**Comment by underfisk** [comment]
@omermorad There's an issue with `rpc-exception` type. Could you add a quick fix so that CI can pass?

**Comment by omermorad** [comment]
> @omermorad There's an issue with `rpc-exception` type. Could you add a quick fix so that CI can pass?

Trying to, I can't find the problem, why would my changes be related to this package?

**Comment by underfisk** [comment]
> > @omermorad There's an issue with `rpc-exception` type. Could you add a quick fix so that CI can pass?
> 
> Trying to, I can't find the problem, why would my changes be related to this package?

It isn't related, all you have to do is add `!` to force the null check or simply add a null coalesce because for some reason this often pop in PR's..

**Comment by underfisk** [comment]
@omermorad I've fixed the issue and added auto-merge so as soon as the checks pass, it should be good to go

**Pull request merged**